File: BIOL6308/Lab9/sliding.py
#!/usr/bin/env python
'''
Name: Dan Shea
Date: 2013-11-08
Description: Compute the GC content of a sequence for a given window size and sequence
'''

import sys

def compute_GC(window_size, sequence):
    '''compute_GC(window_size, sequence): Compute the GC content of a sequence for a given window size and sequence'''
    # Convert sequence to upper case if lower case
    sequence.upper()
    # Calculate the GC content for various windows and return a list with the GC content of that window
    return([(i,(sequence[i:window_size+i].count('G') + sequence[i:window_size+i].count('C'))/float(window_size)) for i in xrange(0,len(sequence)-window_size+1)])

def main():
    if (len(sys.argv) != 3):
        usage()
    else:
        results = compute_GC(int(sys.argv[1]), sys.argv[2])
        for (i,j) in results:
            print '{0:d}\t{1:0.3f}'.format(i,j)
        sys.exit(0)

def usage():
    print 'Usage: {0:s} window_size sequence'.format(sys.argv[0])
    sys.exit(1)

if (__name__ == '__main__'):
    main()
else:
    usage()

File: BIOL6308/Lab9/sliding.pl
#!/usr/bin/env perl
use strict;
use warnings;
use diagnostics;

# Implement a Perl program that will take two arguments in ARGV: The first argument is the window size, and the second arugment
# is the sequence. This program will slide over the sequence calculating the gc content at each window size. You should be able
# to run the Perl program like:
#
# perl sliding.pl 5 ACGATAGAAAATGCTAAG
#
# And the output should look like:
#
# 0	0.400
# 1	0.400
# 2	0.400
# 3	0.200
# 4	0.200
# 5	0.200
# 6	0.200
# 7	0.000
# 8	0.200
# 9	0.400
# 10	0.400
# 11	0.400
# 12	0.400
# 13	0.400
#
# HINT: Use a nested for loop, the substr function, and the printf function to get nicely rounded numbers

my $window_size = $ARGV[0];
my $sequence = $ARGV[1];

for (my $i=0; $i < length($sequence)-$window_size-1; $i++){
    my $count = () = substr($sequence, $i, $window_size) =~ /[GC]/gi;
    printf("%d\t%0.3f\n",$i, $count/$window_size);
}